Feat: lesson20

yuka830 opened the pull request:
# [Lesson20](https://github.com/kenmori/handsonFrontend/blob/master/work/markup/1.md#20)
```
こちらのようなテーブルを画面遷移してから3秒後に解決されるPromiseが返すオブジェクトを元に作ってください。
カラム名など(id, 名前等)もdataで表現して受け取り、フロント側で加工して表示すること
```
## [CodeSandbox](https://codesandbox.io/s/lesson20-kmgu9?file=/main.js)
Last commit: [bc308a9](https://github.com/yuka830/js-lessons/pull/26/commits/bc308a9af505a3c053396b3ddf4ac9ea258a491d)

Implemented the user table using `forEach`. The table header is created from the key in the object.
I was taught don't use the upper case in json file in the past lessons, so this time I used `toUpperCase()` to display for `id` in uppercase.
I'm a little worry about `カラム名など(id, 名前等)もdataで表現して受け取り、フロント側で加工して表示すること`
Did I do in the correct way? I'm sorry if I misunderstood. Thanks!

AyakaOnozuka131 commented:
Putting a JSDoc comment in front of each function makes it easier to read.
It also makes it easier to understand when you look back at it yourself later.
[JavaScriptでJSDocコメントを書くメリットとは](https://ics.media/entry/6789/)

If you are using VScode, the following plug-ins are recommended.
[Document This](https://marketplace.visualstudio.com/items?itemName=oouo-diogo-perdigao.docthis)

ghost commented:
IMO
Try changing the json data to something a little more realistic; the key should not be in kanji when retrieving it via the API.
```js
{
  "id":"1",
  "名前":"やまだ",
  "性別":"男",
  "年齢":"32"
}
```
```js
 {
  "id": 3, //number
   "name": "たなか",
   "sex": "男",
   "age": 32  //number
 }
```

yuka830 commented:
@AyakaOnozuka131 
> Putting a JSDoc comment in front of each function makes it easier to read. It also makes it easier to understand when you look back at it yourself later. [JavaScriptでJSDocコメントを書くメリットとは](https://ics.media/entry/6789/)
> 
> If you are using VScode, the following plug-ins are recommended. [Document This](https://marketplace.visualstudio.com/items?itemName=oouo-diogo-perdigao.docthis)

Thank you for teaching me about this! It's new to me :)
I'm still not sure how to use it well, so I will search more. Please let me ask you if I got some questions.

yuka830 commented:
@443n0511
> IMO Try changing the json data to something a little more realistic; the key should not be in kanji when retrieving it via the API.

Thanks! I recreated the json file separately for the title and user data. Also fixed the problem of numbers being strings.
[Newest JSON file](https://myjson.dit.upm.es/api/bins/dx91)
First fix: [752bc66](https://github.com/yuka830/js-lessons/pull/26/commits/752bc665e104a0cee6aa77307fd36823987929c7)
Second fix: [dbb0089](https://github.com/yuka830/js-lessons/pull/26/commits/dbb00890d9cccf42526ed3e38464493a8b5623be)
(I recreated JSON file again. Changed `title` of JSON data from object to array )

yuka830 commented:
@AyakaOnozuka131 
I tried writing JSDoc reading your reference url. It said "unless obvious from the property, method, or parameter name." in Google JavaScript Style Guide, so I wrote it just for some of them.
How about them? I'm not sure it's correct. If I'm wrong please kindly teach me. Thanks!
[ab68607](https://github.com/yuka830/js-lessons/pull/26/commits/ab68607fdb10f4585eb42d46edd985a75131dd83)

ghost commented:
When I took on this challenge, I also created json for the title data and content data separately.
But…
Mr. Morita taught us that we are too dependent on the client side.
I'll transcribe his comments here.
> This API value is unusual.
The value of title is too client side view dependent.
Remove title object. like below.
You should be able to make this value.
```json
{
    "data": [


        {
            "id": 1,
            "name": "やまだ",
            "sex": "男",
            "age": 32
        },
        {
            "id": 2,
            "name": "さとう",
            "sex": "女",
            "age": 18
        },
        {
            "id": 3,
            "name": "たなか",
            "sex": "男",
            "age": 25
        },
        {
            "id": 4,
            "name": "あんどう",
            "sex": "女",
            "age": 48
        },

        {
            "id": 5,
            "name": "えのもと",
            "sex": "女",
            "age": 36
        }
    ]
}
```
Should be able to implement it with just this info! Good luck!

yuka830 commented:
@443n0511 
> When I took on this challenge, I also created json for the title data and content data separately.
But…
Mr. Morita taught us that we are too dependent on the client side.
I'll transcribe his comments here.

Thank you for teaching me. I understand about this.
I've changed JSON file for the one you told me and reimplemented some of function. 
For the title of table header, I implemented `formingTableHeaderNameWithKey` function with Switch statement.
How about this way?
[3d608e4](https://github.com/yuka830/js-lessons/pull/26/commits/3d608e4154c1b215f5352d60f6a0548f1e80d812)

ghost commented:
@yuka830 
> For the title of table header, I implemented formingTableHeaderNameWithKey function with Switch statement.
How about this way?

Wonderful!
It's amazing what you can implement in one go. I was surprised that you wrote exactly what I intended. Excellent!

443n0511 commented on lesson20/main.js:
```diff
@@ -1 +1,146 @@
+ *Creating table header with key of usersData
+ * @param {Array} usersData The Array of usersData
+ */
+ const createTableHeader = (usersData) => {
+  const trOfThead = document.createElement("tr");
+  const fragment = document.createDocumentFragment();
+  /**
+   *Extract key from usersData
+   * @param {Object} usersData[0] value from one of userData in the object
+   * @param {String} key key of userData
+   */
+  Object.keys(usersData[0]).forEach((key) => {
+    const th = createElementWithClassName("th", "users-table__th");
+    th.textContent = formingTableHeaderNameWithKey(key);
+    fragment.appendChild(th);
+  });
+  trOfThead.appendChild(fragment);
+  return trOfThead;
+};
```
IMO
I think I could cut it out here.
```suggestion
const createTableHeader = (usersData) => {
  const trOfThead = document.createElement("tr");
  const fragment = document.createDocumentFragment();
  AddTextContentToTheads(usersData, fragment);
  trOfThead.appendChild(fragment);
  return trOfThead;
};

/**
 *Extract key from usersData
 * @param {Object} usersData[0] value from one of userData in the object
 * @param {String} key key of userData
 */
const AddTextContentToTheads = (usersData, fragment) => {
  Object.keys(usersData[0]).forEach((key) => {
    const th = createElementWithClassName("th", "users-table__th");
    th.textContent = formingTableHeaderNameWithKey(key);
    fragment.appendChild(th);
  });
};
```

yuka830 replied:
It became more readable! Thank you :)
[2f49e51](https://github.com/yuka830/js-lessons/pull/26/commits/2f49e510284c6b7fe81c680e56de06e94a240df1)

443n0511 commented on lesson20/main.js:
```diff
@@ -1 +1,146 @@
+  return trOfThead;
+};
+
+const formingTableHeaderNameWithKey = (key) => {
+  switch (key) {
+    case "id":
+      return "ID";
+    case "name":
+      return "名前";
+    case "sex":
+      return "性別";
+    case "age":
+      return "年齢";
+    default:
+      console.error();
+  }
+};
```
NITS
I'll tell you why the error happened.
```suggestion
const formingTableHeaderNameWithKey = (key) => {
  switch (key) {
    case "id":
      return "ID";
    case "name":
      return "名前";
    case "sex":
      return "性別";
    case "age":
      return "年齢";
    default:
      console.error(`${key}は見つかりませんでした`);
  }
};
```

yuka830 replied:
I've totally forgot to input about this message 😭 
I'm sorry about that and thank you for telling me.
[e52f0a4](https://github.com/yuka830/js-lessons/pull/26/commits/e52f0a4da0bff0963fa756caf9433c57bf1010eb)

443n0511 commented on lesson20/main.js:
```diff
@@ -1 +1,146 @@
+  loader.classList.add("loaded");
+};
+
+const fetchDataInSecond = (sec, jsonUrl) => {
+  return new Promise((resolve, reject) => {
+    setTimeout(() => {
+      resolve(fetch(jsonUrl));
+    }, sec);
+  });
+};
```
MUST
If we try to make this place like this...
```js
const jsonUrl = "https://myjson.dit.upm.es/api/bins/hogehoge";
```
Nothing was being displayed. Please fix this.
<img width="1088" alt="スクリーンショット 2021-12-31 17 48 01" src="https://user-images.githubusercontent.com/43788628/147813902-c3ddff84-9ef4-494f-b7fd-f8cf37f88961.png">

yuka830 replied:
I've not realized about this.
So I implemented new function which throws error depending on the content of the response, and fixed this problem with this function. 
How about this way?
[5df0264](https://github.com/yuka830/js-lessons/pull/26/commits/5df0264ff494ef8ef49f106d86fcfc70013cb907)
<img width="1343" alt="Screen Shot 2021-12-31 at 13 30 39" src="https://user-images.githubusercontent.com/66966699/147839561-5b344ad5-c3e0-458d-8877-7de125a0d785.png">

443n0511 replied:
Wonderful! It's perfect!

443n0511 commented on lesson20/main.js:
```diff
@@ -1 +1,159 @@
+  return fragment;
+};
+
+const init = async () => {
+  tableWrap.insertAdjacentElement("beforebegin", createLoader());
+  loading();
+  let usersData;
+  try {
+    usersData = await fetchUsersData();
+  } catch (e) {
+    console.error(e);
+  } finally {
+    console.log("処理が完了しました。");
+  }
+  if (usersData.length === 0) {
+    tableWrap.textContent = "データがありません。";
+    return;
+  } else {
+    renderTable(usersData);
+  }
+};
```
IMO
The first one.

FROM
```js
  if (usersData.length === 0) {
    tableWrap.textContent = "データがありません。";
    return;
  } else {
    renderTable(usersData);
  }
```
TO
```js
  if (usersData.length === 0) {
    tableWrap.textContent = "データがありません。";
    return;
  }
  renderTable(usersData);
```
The second.
I think the handling of init() errors needs a little more usability.
<img width="1080" alt="スクリーンショット 2022-01-01 16 00 16" src="https://user-images.githubusercontent.com/43788628/147845581-23514d2e-cd90-4112-a2a1-83000a0245cd.png">
```suggestion
const init = async () => {
  tableWrap.insertAdjacentElement("beforebegin", createLoader());
  loading();
  let usersData;
  try {
    usersData = await fetchほげほげData(); 
  } catch (e) {
  //I feel that if we fix this, it will be even better.
    loaded();
    console.error(e);
    tableWrap.textContent = `error内容:${e.message}`;
  } finally {
    console.log("処理が完了しました。");
  }
  if (usersData.length === 0) {
    tableWrap.textContent = "データがありません。";
    return;
  }
  renderTable(usersData);
};
```

yuka830 replied:
- For the first one
I searched about quick return again and I learned it doesn't need `else`. I didn't understand about that.
Thank you for teaching me :)
[88e2fb9](https://github.com/yuka830/js-lessons/pull/26/commits/88e2fb986e6d3b0b3e4c0ce81f0ea593129c2ae1)

- For the second one
You are right, this improves the usability.
I implemented error messages are displayed in catch statement. I wrote error message in a similar way to other error messages.
And also I change calling `loaded()`　in finally statement not in `fetchJson`  but in `init`. I think finally statement is not must, so I deleted it in `fetchJson`.
What do you think? Thanks reviewing many times.
[bc308a9](https://github.com/yuka830/js-lessons/pull/26/commits/bc308a9af505a3c053396b3ddf4ac9ea258a491d)
